cmd/metalctl/app/move.go

```diff
@@ -0,0 +1,347 @@
+			continue
+		}
+		return nil, fmt.Errorf("a CR %s/%s already exists in the target cluster and is different then in the source cluster", sourceCr.GetNamespace(), sourceCr.GetName())
+	}
+	return crsToMove, nil
+}
+
+type Node struct {
+	Cr       *unstructured.Unstructured
+	Children []*Node
+}
+
+func crsOwnerReferenceTrees(crs []*unstructured.Unstructured) []*Node {
+	nodeMap := make(map[types.UID]*Node)
+
+	for _, cr := range crs {
+		nodeMap[cr.GetUID()] = &Node{Cr: cr}
+	}
+	roots := []*Node{}
+	for _, cr := range crs {
+		if len(cr.GetOwnerReferences()) == 0 || nodeMap[cr.GetOwnerReferences()[0].UID] == nil {
+			roots = append(roots, nodeMap[cr.GetUID()])
+		} else {
+			owner := nodeMap[cr.GetOwnerReferences()[0].UID]
+			owner.Children = append(owner.Children, nodeMap[cr.GetUID()])
+		}
+	}
+	return roots
+}
+
+func cleanup[T client.Object](ctx context.Context, cl client.Client, objs []T) error {
+	cleanupErrs := make([]error, 0)
+	for _, obj := range objs {
+		if err := cl.Delete(ctx, obj); err != nil {
+			cleanupErrs = append(cleanupErrs, err)
+		}
+	}
+	return errors.Join(cleanupErrs...)
+}
+
+func moveCrds(ctx context.Context, cl client.Client, sourceCrds []*apiextensionsv1.CustomResourceDefinition) (movedCrds []*apiextensionsv1.CustomResourceDefinition, err error) {
+	movedCrds = make([]*apiextensionsv1.CustomResourceDefinition, 0)
+	for _, crd := range sourceCrds {
+		crd.SetResourceVersion("")
+		if err = cl.Create(ctx, crd); err != nil {
+			err = fmt.Errorf("CRD %s couldn't be created in the target cluster: %w", crdKind(crd), err)
+			return
+		}
+		movedCrds = append(movedCrds, crd)
+	}
+
+	// wait for CRDs to be present on the target cluster
+	err = wait.PollUntilContextTimeout(ctx, time.Second, 30*time.Second, true, func(ctx context.Context) (bool, error) {
+		for _, crd := range movedCrds {
+			targetObj := apiextensionsv1.CustomResourceDefinition{}
+			err := cl.Get(ctx, client.ObjectKeyFromObject(crd), &targetObj)
+			if err != nil {
+				return false, client.IgnoreNotFound(err)
+			}
+		}
+		return true, nil
+	})
+
+	return
+}
+
+func moveCrs(ctx context.Context, cl client.Client, crsTrees []*Node, uid ...types.UID) (movedCrs []*unstructured.Unstructured, err error) {
+	movedCrs = make([]*unstructured.Unstructured, 0)
+
+	for _, crsTree := range crsTrees {
+		ownerReferences := crsTree.Cr.GetOwnerReferences()
+		if len(ownerReferences) == 1 && len(uid) == 1 {
+			ownerReferences[0].UID = uid[0]
+			crsTree.Cr.SetOwnerReferences(ownerReferences)
+		}
+
+		crsTree.Cr.SetResourceVersion("")
+		if err = cl.Create(ctx, crsTree.Cr); err != nil {
+			err = fmt.Errorf("CR %s couldn't be created in the target cluster: %w", crName(crsTree.Cr), err)
+			return
+		}
+		movedCrs = append(movedCrs, crsTree.Cr)
+	}
+
+	for _, crsTree := range crsTrees {
+		err = wait.PollUntilContextTimeout(ctx, time.Second, 30*time.Second, true, func(ctx context.Context) (bool, error) {
```

Comment:
What is the intent of the wait loop?

Reply:
When creating children CRs they should have set owner reference. So optional argument `uid` is passed to `moveCrs` function. It is empty when creating `root` CRs (without owner or with owner not in metal CRs). It should be set when creating children CRs so owner CR must be created on the target cluster and its UID must be passed. This is why there is a wait loop, to retrive uid of an owner.

Reply:
After a CR is created it takes some time to get it from api, so `Get` method might return an error at first.